src/Shared/Hash.cs

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
```

Comment:
We have other code using this combiner: https://github.com/aspnet/Common/tree/dev/src/Microsoft.Extensions.HashCodeCombiner.Sources

I don't mind which we use, but we should consolidate on one :). Doesn't have to happen in the PR, but might be worth tracking for clean-up later.